### PR TITLE
Export: allow student to upload icon image for exported app

### PR DESCRIPTION
**Pull request by cpirich**
* Completed the "Upload your App icon" screen in the Export flow - as described in this [spec](https://docs.google.com/document/d/1o-C8GOi45FCzUKPRnsyrNS8k10jUYOlrP2zIBPeO1X4/edit)
* Per discussion with @ryansloan, when the student uploads an icon, we will use it as both the icon and the splash screen. Verified by installing an APK on Android that this works.

Screen Shot 2019-07-19 at 6 42 26 PM

Launching the Android app:

ezgif-4-a629a531ab8c

### CODE REVIEW

File: apps/src/code-studio/components/ExportDialog/IconPage.jsx [hunk 1]
```diff
@@ -54,11 +55,26 @@ export default class IconPage extends React.Component {
...
+  handleSelectUploadFile = () => {
+    if (!this.uploadFileInput.value) {
+      return;
+    }
+    const {onIconSelected} = this.props;
+    onIconSelected(URL.createObjectURL(this.uploadFileInput.files[0]));
+  };
+
+  onUploadIcon = () => {
+    this.uploadFileInput.click();
```
Comment (jmkulwik) [inline apps/src/code-studio/components/ExportDialog/IconPage.jsx:78]: Why not just override the input onClick event (and value)? This seems overly complicated.

File: apps/src/code-studio/components/ExportDialog/IconPage.jsx [hunk 2]
```diff
@@ -54,11 +55,26 @@ export default class IconPage extends React.Component {
...
+          accept="image/png"
+          onChange={this.handleSelectUploadFile}
+        />
       </div>
     );
   }
 
-  uploadIcon = () => {
-    // TODO: implement icon upload
+  handleSelectUploadFile = () => {
```
Comment (jmkulwik) [inline apps/src/code-studio/components/ExportDialog/IconPage.jsx]: Maybe I just need another ☕️, but it seems like there are too many ways to interpret this function name. Mind renaming or commenting? (i.e. It can be interpreted as `selectFile`, `uploadFile`, `selectFileToUpload` or `selectedUploadFileFromList`)

Comment (cpirich) [reply apps/src/code-studio/components/ExportDialog/IconPage.jsx]: This was adapted from another piece of code in our codebase. I agree that the name is confusing

I think I'll change it to the more obvious `onUploadFileInputChange` (since we're handling `onChange` for what we are calling the `uploadFileInput`)

File: apps/src/code-studio/components/ExportDialog/IconPage.jsx (hunk 1 above)
Comment (cpirich) [reply apps/src/code-studio/components/ExportDialog/IconPage.jsx:78]: The `input` element is hidden, so you can't click on that. We want to show a button (not an input). So the "trick" is to have the button press route a click to the hidden `input`.

If you search for `type="file"` in the codebase, you'll see a few other places that do the same thing.

Comment (joshlory) [reply apps/src/code-studio/components/ExportDialog/IconPage.jsx:78]: Can we generalize that pattern as a component? (It looks like we already have with [`<HiddenUploader/>`](https://github.com/code-dot-org/code-dot-org/blob/1cad5ac667f9ed1a3aa31063e6d06bd21a78e2bd/apps/src/code-studio/components/HiddenUploader.jsx)?)

Comment (cpirich) [reply apps/src/code-studio/components/ExportDialog/IconPage.jsx:78]: @joshlory I looked at `HiddenUploader` and it has all sorts of baggage. It uses `jquery` and assumes we want to upload to a URL. In this case, we're not actually uploading anything - we're just trying to grab a `File` object and convert it to an Object URL. The `snack-sdk` will be used for the uploading.

Comment (joshlory) [reply apps/src/code-studio/components/ExportDialog/IconPage.jsx:78]: Ah I see. Maybe there's an inner component we can extract? Either way, thanks for checking!

Comment (cpirich) [reply apps/src/code-studio/components/ExportDialog/IconPage.jsx:78]: For future: the common piece I could see us extracting could be a tiny React wrapper around the `<input type="file">` with a bunch of props to handle the various cases. If we wanted to re-create `HiddenUploader` on top of that component, we could.

Comment (jmkulwik) [reply apps/src/code-studio/components/ExportDialog/IconPage.jsx:78]: Chris and I had an offline conversation about this as well, and I want to capture it here as it's valuable for future readers.
___
> Hey, looking at this, would it make sense to replace the button with the input element (and style the input element like a button)? Then we don't need to route between two elements.

- As far as I know, that's not possible. This is the way I've seen it done for a while. I'll do some reading online to see if what you are suggesting has now become possible with newer browsers...
An article from a while ago that tries to do something similar, but they still have a hidden `input` and instead style the associated `label`!: https://tympanus.net/codrops/2015/09/15/styling-customizing-file-inputs-smart-way/

- A newer attempt - but the comments make reference to still needing to trap the `click` event for Internet Explorer: https://coderwall.com/p/uer3ow/total-input-type-file-style-control-with-pure-css

- A 3rd article here that also uses labels, but has some funky CSS that doesn't completely hide the `input` and references a Firefox bug: https://benmarshall.me/styling-file-inputs/

- Long story short, I think we're better off sticking with what has been working in our project up until now, even though I agree that it's an odd aspect of HTML that we're trying to workaround.
___

Comment (jmkulwik) [reply apps/src/code-studio/components/ExportDialog/IconPage.jsx:78]: @cpirich, would you be able to make the react wrapper you mentioned in your earlier message as part of this change or as a follow-up change?

Comment (cpirich) [reply apps/src/code-studio/components/ExportDialog/IconPage.jsx:78]: @jmkulwik I don't think we should do it as part of this PR for two reasons:
* This scope of this PR will get a little too large and include a re-factor that impacts all users of `HiddenUploader` (seems better to do with a separate ticket/PR)
* There is some additional risk once we change how `HiddenUploader` works that will require more thoughtful testing to be sure we got it right.

**Review by jmkulwik — APPROVED** [review]
Thanks Chris!
